### PR TITLE
Implements the undersampling of kspace z direction

### PR DESCRIPTION
**Pull request by mathrip**
This PR implements the undersampling of the k-space in the z-direction (only working for the stacks implementations).

This solves #56 as the z-dimension has been cropped, it also solves #33

### CODE REVIEW

File: mri/operators/fourier/utils.py [hunk 1]
```diff
@@ -257,7 +257,8 @@ def get_stacks_fourier(kspace_loc):
     ----------
     ksapce_plane_loc: np.ndarray
```
Comment (philouc) [inline mri/operators/fourier/utils.py]: typo: ksapce -> kspace

File: mri/operators/fourier/utils.py [hunk 2]
```diff
@@ -257,7 +257,8 @@ def get_stacks_fourier(kspace_loc):
...
         the mask samples in the 3D Fourier domain.
-
+    shape: tuple
+        Reconstructed volume shape
     Returns
     ----------
     ksapce_plane_loc: np.ndarray
```
Comment (philouc) [inline mri/operators/fourier/utils.py]: same typo

File: mri/operators/fourier/utils.py [hunk 3]
```diff
@@ -266,15 +267,26 @@ def get_stacks_fourier(kspace_loc):
         A 1D array of z-sample locations
     sort_pos: np.ndarray
         The sorting positions for opertor and inverse for incoming data
+    idx_mask_z: np.ndarray
+        contains the indices of the acquired Fourier plane
```
Comment (philouc) [inline mri/operators/fourier/utils.py]: Fourier planes instead of plane?

File: mri/operators/fourier/utils.py (hunk 1 above)
Comment (philouc) [inline mri/operators/fourier/utils.py]: update name variable given the typo

File: mri/operators/fourier/utils.py [hunk 4]
```diff
@@ -288,4 +300,4 @@ def get_stacks_fourier(kspace_loc):
     ksapce_plane_loc = stacked[0, :, 0:2]
     z_sample_loc = stacked[:, 0, 2]
     z_sample_loc = z_sample_loc[:, np.newaxis]
-    return ksapce_plane_loc, z_sample_loc, sort_pos
+    return ksapce_plane_loc, z_sample_loc, sort_pos, idx_mask_z
```
Comment (philouc) [inline mri/operators/fourier/utils.py]: propagate changes on name variable

File: mri/operators/fourier/utils.py [hunk 5]
```diff
@@ -255,37 +255,53 @@ def get_stacks_fourier(kspace_loc):
...
     kspace_loc = kspace_loc[sort_pos]
-    first_stack_len = np.size(np.where(
-        kspace_loc[:, 2] == np.min(kspace_loc[:, 2])))
+
+    # Find the mask used to sample stacks in z direction
+    full_stack_loc = convert_mask_to_locations(np.ones(shape[2]))[:, 0]
+    sampled_stack_loc = np.unique(kspace_loc[:, 2])
+
+    try:
+        idx_mask_z = np.asarray([np.where(x == full_stack_loc)[0][0] for
```
Comment (zaccharieramzi) [inline mri/operators/fourier/utils.py]: It feels like this doesn't really belong to the function. It's doing something completely parallel to the rest of it. Why does it need to be in this function ?

File: mri/operators/fourier/utils.py [hunk 6]
```diff
@@ -255,37 +255,53 @@ def get_stacks_fourier(kspace_loc):
...
+                                       np.min(kspace_loc[:, 2])))
     acq_num_slices = int(len(kspace_loc) / first_stack_len)
+
     stacked = np.reshape(kspace_loc, (acq_num_slices,
                                       first_stack_len, 3))
     z_expected_stacked = np.reshape(np.repeat(stacked[:, 0, 2],
                                               first_stack_len),
                                     (acq_num_slices,
                                      first_stack_len))
+
```
Comment (zaccharieramzi) [inline mri/operators/fourier/utils.py]: Let's try to avoid as much as possible whiteline changes. They mess up the git history and clutter the PR.

File: mri/parallel_mri/extract_sensitivity_maps.py [hunk 7]
```diff
@@ -94,25 +94,30 @@ def gridded_inverse_fourier_transform_nd(kspace_loc,
...
-        This holds the z-sample locations for stacks. Again, extracted using
+    idx_mask_z: np.ndarray
+        contains the indices of the acquired Fourier plane. Extracted using
         get_stacks_fourier function
-    kspace_data: np.ndarray
-        The k-space data corresponding to kspace_plane_loc above
-    grid: np.ndarray
+    grid: tuple
         The Gridded matrix for which you want to calculate k_space Smaps
+    shape: tuple
```
Comment (zaccharieramzi) [inline mri/parallel_mri/extract_sensitivity_maps.py]: same, maybe `volume_shape` is more adapted.

File: mri/parallel_mri/extract_sensitivity_maps.py [hunk 8]
```diff
@@ -121,17 +126,16 @@ def gridded_inverse_fourier_transform_stack(kspace_plane_loc, z_sample_loc,
...
-                     fill_value=0))
-    # Move the slice axis to last : Make to Nx x Ny x Nz
-    gridded_kspace = np.moveaxis(np.asarray(gridded_kspace), 0, 2)
+    for i, idx_z in enumerate(idx_mask_z):
+        gridded_kspace[:, :, idx_z] = griddata(
+            kspace_plane_loc,
+            kspace_data_sorted[i*stack_len:(i+1)*stack_len],
+            grid,
+            method=method,
+            fill_value=0)
```
Comment (zaccharieramzi) [inline mri/parallel_mri/extract_sensitivity_maps.py]: nitpicking here, but to handle multi-line in the best way possible you can even do:
```python
gridded_kspace[:, :, idx_z] = griddata(
  kspace_plane_loc,
  kspace_data_sorted[i*stack_len:(i+1)*stack_len],
  grid,  
  method=method,
  fill_value=0,
)
```
If for example someone wants to add an argument at the end, they don't need to modify the `fill_value` line too.

File: mri/parallel_mri/extract_sensitivity_maps.py [hunk 9]
```diff
@@ -215,8 +219,8 @@ def get_Smaps(k_space, img_shape, samples, thresh,
                                   endpoint=False)
                       for i in np.arange(np.size(img_shape)-1)]
         grid = np.meshgrid(*grid_space)
-        kspace_plane_loc, z_sample_loc, sort_pos = \
-            get_stacks_fourier(samples)
+        kspace_plane_loc, _, sort_pos, idx_mask_z = \
```
Comment (zaccharieramzi) [inline mri/parallel_mri/extract_sensitivity_maps.py]: Then do we need to have `z_sample_loc` returned in the `get_stacks_fourier` function?

File: mri/parallel_mri/extract_sensitivity_maps.py [hunk 10]
```diff
@@ -94,25 +94,30 @@ def gridded_inverse_fourier_transform_nd(kspace_loc,
...
         in plane, extracted using get_stacks_fourier function
-    z_sample_loc: np.ndarray
-        This holds the z-sample locations for stacks. Again, extracted using
+    idx_mask_z: np.ndarray
+        contains the indices of the acquired Fourier plane. Extracted using
         get_stacks_fourier function
-    kspace_data: np.ndarray
-        The k-space data corresponding to kspace_plane_loc above
-    grid: np.ndarray
+    grid: tuple
```
Comment (zaccharieramzi) [inline mri/parallel_mri/extract_sensitivity_maps.py]: why did the type change?

File: mri/operators/fourier/non_cartesian.py [hunk 11]
```diff
@@ -482,8 +482,12 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
         """
         self.num_slices = shape[2]
         self.shape = shape
-        (kspace_plane_loc, self.z_sample_loc,
-         self.sort_pos) = get_stacks_fourier(kspace_loc)
+        (kspace_plane_loc,
+         self.z_sample_loc,
+         self.sort_pos,
+         self.idx_mask_z) = get_stacks_fourier(
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: I think this can be on a single line (it's just 91 characters)

File: mri/operators/fourier/non_cartesian.py [hunk 12]
```diff
@@ -494,18 +498,25 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
     def _op(self, data):
         fft_along_z_axis = np.fft.fftshift(np.fft.fft(
             np.fft.ifftshift(data, axes=2),
-            n=self.acq_num_slices,
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: why was this removed?

File: mri/operators/fourier/non_cartesian.py [hunk 13]
```diff
@@ -494,18 +498,25 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
...
             norm="ortho"),
             axes=2)
+
         stacked_kspace = np.asarray(
             [self.plane_fourier_operator.op(fft_along_z_axis[:, :, slice])
-             for slice in np.arange(self.acq_num_slices)])
-        stacked_kspace = np.reshape(stacked_kspace,
-                                    self.acq_num_slices * self.stack_len)
+             for slice in range(self.num_slices)])
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: while we are changing this line, let's use a different variable name, `slice` is protected in Python.

File: mri/operators/fourier/non_cartesian.py [hunk 14]
```diff
@@ -494,18 +498,25 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
...
+        stacked_kspace_sampled = np.zeros((self.acq_num_slices,
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: I think this is legacy

File: mri/operators/fourier/non_cartesian.py [hunk 15]
```diff
@@ -494,18 +498,25 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
...
+                                           self.stack_len),
+                                          dtype=data.dtype)
+        stacked_kspace_sampled = stacked_kspace[self.idx_mask_z, :]
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: You can just do `stacked_kspace_sampled = stacked_kspace[self.idx_mask_z]`

File: mri/operators/fourier/non_cartesian.py [hunk 16]
```diff
@@ -494,18 +498,25 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
...
+
+        stacked_kspace_sampled = np.reshape(
+            stacked_kspace_sampled,
+            self.acq_num_slices * self.stack_len)
         # Unsort the Coefficients
         inv_idx = np.zeros_like(self.sort_pos)
         inv_idx[self.sort_pos] = np.arange(len(self.sort_pos))
-        return stacked_kspace[inv_idx]
+        return stacked_kspace_sampled[inv_idx] * \
+            np.sqrt(self.num_slices / self.acq_num_slices)
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: Can you add a comment on this normalisation?

File: mri/operators/fourier/test_fourier_adjoint.py [hunk 17]
```diff
@@ -196,8 +196,14 @@ def test_similarity_stack_3D(self):
...
                 _mask = np.random.randint(2, size=(N, N))
-                _mask3D = np.asarray([_mask for i in np.arange(Nz)])
-                _samples = convert_mask_to_locations(_mask3D.swapaxes(0, 2))
+
+                # Generate random mask along z
+                sampling_z = np.random.randint(2, size=Nz)
+                _mask3D = np.zeros((N, N, Nz))
+                for idx, acq_z in enumerate(sampling_z):
+                    _mask3D[:, :, idx] = _mask * acq_z
+                _samples = convert_mask_to_locations(_mask3D)
```
Comment (zaccharieramzi) [inline mri/operators/fourier/test_fourier_adjoint.py]: I can see that they were here before, but I don't understand why these variable names begin with `_`.

File: mri/operators/fourier/utils.py (hunk 5 above)
Comment (mathrip) [reply mri/operators/fourier/utils.py]: This function returns the characteristics of the k-space, the mask of stacks in z-direction is one of its characteristics. I think this need to be returned by this function.

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 10 above)
Comment (mathrip) [reply mri/parallel_mri/extract_sensitivity_maps.py]: idx_mask_z is needed. z_sample_loc is redondant as we already have the kspace_data to give the z-locations.

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 9 above)
Comment (mathrip) [reply mri/parallel_mri/extract_sensitivity_maps.py]: z_sample_loc is requiered in other call of the function get_stacks_fourier

File: mri/operators/fourier/utils.py [hunk 18]
```diff
@@ -245,7 +245,7 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
     return gradient_op, linear_op, prox_op, cost_op
 
 
-def get_stacks_fourier(kspace_loc):
+def get_stacks_fourier(kspace_loc, shape):
```
[the remark this answers is not in the document]
Comment (mathrip) [reply mri/operators/fourier/utils.py]: I agree, however, there is a lot of dependencies for this function, and a change of variable name would lead to change multiples calls in the examples and test scripts. It is best if it stays as "shape" for now.

File: mri/operators/fourier/utils.py [hunk 19]
```diff
@@ -255,37 +255,53 @@ def get_stacks_fourier(kspace_loc):
...
     Returns
     ----------
-    ksapce_plane_loc: np.ndarray
+    kspace_plane_loc: np.ndarray
         A 2D array of samples which when stacked gives the 3D samples
     z_sample_loc: np.ndarray
         A 1D array of z-sample locations
     sort_pos: np.ndarray
         The sorting positions for opertor and inverse for incoming data
+    idx_mask_z: np.ndarray
```
[the remark this answers is not in the document]
Comment (LElgueddari) [reply mri/operators/fourier/utils.py]: What do you mean by masking entire planes?
Masking actually means not acquiring so the doc should mention that it's possible to not acquire entire plane, is that what you mean?

File: mri/operators/fourier/utils.py [hunk 20]
```diff
@@ -248,15 +248,16 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
...
-    checks for any issues of the incoming k-space pattern and if the stack
-    property is not satisfied.
+    values of the stacks acquiered and converts to stacks of 2D.
+    This function also checks for any issues of the incoming k-space
+    pattern and if the stack property is not satisfied.
     Stack Property: The k-space locations originate from a stack of 2D samples.
 
     Parameters
     ----------
     kspace_plane_loc: np.ndarray
```
Comment (LElgueddari) [inline mri/operators/fourier/utils.py]: The parameter name is `kspace_loc` not `kspace_plane_loc` which actually corresponds to the acquired k-space samples position in the 3 dimension.

File: mri/operators/fourier/utils.py [hunk 21]
```diff
@@ -248,15 +248,16 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
...
-        the mask samples in the 3D Fourier domain.
+        the mask samples in the 3D Fourier domain
+        possibility to mask entire planes (sampling in z direction)
```
Comment (LElgueddari) [inline mri/operators/fourier/utils.py]: I'm sorry but this doesn't make any sense to me

File: mri/parallel_mri/extract_sensitivity_maps.py [hunk 22]
```diff
@@ -134,7 +134,8 @@ def gridded_inverse_fourier_transform_stack(kspace_data_sorted,
             kspace_data_sorted[i*stack_len:(i+1)*stack_len],
             grid,
             method=method,
-            fill_value=0)
+            fill_value=0
```
Comment (LElgueddari) [inline mri/parallel_mri/extract_sensitivity_maps.py]: I think that what @zaccharieramzi  asked is to add a comma after `fill_value=0`, otherwise there is no point of doing this

File: examples/non_cartesian_reconstruction_stack_3d.py [hunk 23]
```diff
@@ -60,15 +74,20 @@
...
-kspace_plane_loc, z_sample_loc, sort_pos = get_stacks_fourier(kspace_loc)
-grid_soln = gridded_inverse_fourier_transform_stack(kspace_plane_loc,
-                                                    z_sample_loc,
-                                                    kspace_obs,
-                                                    tuple(grid),
-                                                    'linear')
+kspace_plane_loc, z_sample_loc, sort_pos, idx_mask_z = get_stacks_fourier(
+    kspace_loc,
+    image.shape)
+grid_soln = gridded_inverse_fourier_transform_stack(
```
Comment (chaithyagr) [inline examples/non_cartesian_reconstruction_stack_3d.py]: Ahh, this is bad, we would want the stacking done by the function itself. Can we have an issue for this?
We can rename this function and write another function that actually does gridded_inverse_fourier transform that takes in just the stack.

File: mri/operators/fourier/non_cartesian.py (hunk 11 above)
Comment (chaithyagr) [reply mri/operators/fourier/non_cartesian.py]: Would this be better?
```python
(kspace_plane_loc, self.z_sample_loc, self.sort_pos, self.idx_mask_z) \
            = \
            get_stacks_fourier(
                kspace_loc,
                self.shape,
            )
```

I think @zaccharieramzi  is best to suggest the right way here, but to me the above is more readable.

File: mri/operators/fourier/non_cartesian.py (hunk 14 above)
Comment (chaithyagr) [reply mri/operators/fourier/non_cartesian.py]: Yes, we can remove this based on new codes, this seems like a redundant line here

File: mri/operators/fourier/non_cartesian.py (hunk 12 above)
Comment (chaithyagr) [reply mri/operators/fourier/non_cartesian.py]: @zaccharieramzi this is removed to have a full FFT, they will use this to later choose what undersampled data they want.

File: mri/operators/fourier/non_cartesian.py [hunk 24]
```diff
@@ -494,18 +498,25 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
...
-            [self.plane_fourier_operator.op(fft_along_z_axis[:, :, slice])
-             for slice in np.arange(self.acq_num_slices)])
-        stacked_kspace = np.reshape(stacked_kspace,
-                                    self.acq_num_slices * self.stack_len)
+            [self.plane_fourier_operator.op(fft_along_z_axis[:, :, stack])
+             for stack in range(self.num_slices)])
+        stacked_kspace_sampled = np.zeros((self.acq_num_slices,
+                                           self.stack_len),
+                                          dtype=data.dtype)
+        stacked_kspace_sampled = stacked_kspace[self.idx_mask_z]
```
Comment (chaithyagr) [inline mri/operators/fourier/non_cartesian.py]: Please correct me if my understanding here is wrong:
You are effectively doing a complete 2D NFFT on each slice location and choosing undersampled data.
Can we move to just do NFFT on only the slices where we want?

Something like:

```python
stacked_kspace_sampled = np.asarray(
            [self.plane_fourier_operator.op(fft_along_z_axis[:, :, stack])
             for stack in self.idx_mask_z])
```

Was this explored and would this cause any issues?

File: mri/operators/fourier/non_cartesian.py [hunk 25]
```diff
@@ -526,29 +537,26 @@ def op(self, data):
             coeff = [self._op(data[i])
                      for i in range(self.n_coils)]
         coeff = np.asarray(coeff)
+
         return coeff
 
     def _adj_op(self, coeff):
         coeff = coeff[self.sort_pos]
         stacks = np.reshape(coeff, (self.acq_num_slices, self.stack_len))
         # Receive First Fourier transformed data (per plane) in Nz x N x N
```
Comment (chaithyagr) [inline mri/operators/fourier/non_cartesian.py]: It looks like based on code update, you are organizing data as `N x N x Nz`. Can you please update this comment?
Also, it would help to mention this in doc if not done already.

File: mri/operators/fourier/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: You can use refactor in pycharm for this, its good practice to use the right names from start, can we please get this?

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 9 above)
Comment (chaithyagr) [reply mri/parallel_mri/extract_sensitivity_maps.py]: They both represent the same thing after all right? One is z_locations and other just indexes to these locations, can we open an issue to clean this function up? It would be great to get in this PR also, but that's your call..

File: examples/non_cartesian_reconstruction_stack_3d.py (hunk 23 above)
Comment (mathrip) [reply examples/non_cartesian_reconstruction_stack_3d.py]: done !

Comment (mathrip) [reply examples/non_cartesian_reconstruction_stack_3d.py]: done

File: mri/operators/fourier/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mathrip) [reply mri/operators/fourier/utils.py]: I know, the problem is that it will affect some functions and scripts not related to this PR. I think this is a typo problem that needs to be fixed in a special PR to improve the code, make it user-friendly. I raised an issue #61

File: mri/operators/fourier/non_cartesian.py (hunk 13 above)
Comment (mathrip) [reply mri/operators/fourier/non_cartesian.py]: done !

File: mri/operators/fourier/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: I am sorry but I dont understand. We are not referring to all the refactoring mentioned in #61 . This PR adds new parameter shape. Now we would prefer if the code enters with right nomenclature directly. Other name changes can be done in another PR. But this particular one looks to me like is added in this PR and it only makes sense to add the right code in first commit than fix it up later.

[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: I am sorry but I dont understand. We are not referring to all the refactoring mentioned in #61 . This PR adds new parameter shape. Now we would prefer if the code enters with right nomenclature directly. Other name changes can be done in another PR. But this particular one looks to me like is added in this PR and it only makes sense to add the right code in first commit than fix it up later.

File: mri/operators/fourier/non_cartesian.py (hunk 24 above)
Comment (mathrip) [reply mri/operators/fourier/non_cartesian.py]: I made the change, thank you

File: mri/operators/fourier/non_cartesian.py (hunk 25 above)
Comment (mathrip) [reply mri/operators/fourier/non_cartesian.py]: Comment updated. There is already mention of the shape N x N x Nz in the function doc, I don't touch it.

File: mri/operators/fourier/utils.py (hunk 6 above)
Comment (mathrip) [reply mri/operators/fourier/utils.py]: whiteline removed

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 9 above)
Comment (mathrip) [reply mri/parallel_mri/extract_sensitivity_maps.py]: I raised an issue and will try to work on that for this PR

File: mri/operators/fourier/utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: Yes, from what I understand now we have the possibility to undersample in the stack direction as well right?

File: mri/operators/fourier/utils.py (hunk 6 above)
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: I don't see it removed

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 10 above)
Comment (zaccharieramzi) [reply mri/parallel_mri/extract_sensitivity_maps.py]: No I'm asking why the type of `grid` changed

File: mri/operators/fourier/utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (LElgueddari) [reply mri/operators/fourier/utils.py]: Yes we do. It is just the term mask that do not make that much sense. We do not really mask the 3D k-space since each plane is non-Cartesian therefore it cannot be considered as masked Fourier space. 
Beacause of this specific point the following documentation is false
https://github.com/CEA-COSMIC/pysap-mri/blob/cf4dff5dcb28058d946fbf210e584ededa727ec8/mri/operators/fourier/utils.py#L258-L259
It should be replaced by 
```
Acquired 3D k-space locations such that the each plane is composed of the same non-Cartesian samples, 
while the third dimension correspond to a Cartesian under-sampling
```
Something like this

[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: I see !

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 10 above)
Comment (mathrip) [reply mri/parallel_mri/extract_sensitivity_maps.py]: Because we realised it is a tuple (containing two arrays). 
Do you prefere if I precise tuple of np.array?

File: mri/operators/fourier/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: @mathrip, I am not sure I understand your point. You introduced this variable so it means that everywhere the keyword argument `shape` is used for this function is because you introduced it, or am I mistaken?

File: mri/operators/fourier/utils.py [hunk 26]
```diff
@@ -245,33 +245,50 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
...
+        contains the indices of the acquired Fourier planes (z direction)
     """
     # Sort the incoming data based on Z, Y then X coordinates
     # This is done for easier stacking
     sort_pos = np.lexsort(tuple(kspace_loc[:, i]
                                 for i in np.arange(3)))
     kspace_loc = kspace_loc[sort_pos]
+
+    # Find the mask used to sample stacks in z direction
+    full_stack_z_loc = \
```
Comment (zaccharieramzi) [inline mri/operators/fourier/utils.py]: I think it's preferable to limit the use of `\`, you could do:
```python
full_stack_z_loc = convert_mask_to_locations(
  np.ones(volume_shape[2]),
)[:, 0]

File: mri/operators/fourier/utils.py [hunk 27]
```diff
@@ -245,33 +245,50 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
...
+        convert_mask_to_locations(np.ones(volume_shape[2]))[:, 0]
+    sampled_stack_z_loc = np.unique(kspace_loc[:, 2])
+
+    try:
```
Comment (zaccharieramzi) [inline mri/operators/fourier/utils.py]: I think it would be nice to include a comment saying what exactly the test is here.
I understood the first time, and already forgot it...

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 10 above)
Comment (zaccharieramzi) [reply mri/parallel_mri/extract_sensitivity_maps.py]: ok the grid is expressed as a tuple of `ndarray`s?
Maybe not in the type, but in the description of the variable it would be nice.

File: mri/operators/fourier/non_cartesian.py [hunk 28]
```diff
@@ -482,8 +482,12 @@ def __init__(self, kspace_loc, shape, implementation='cpu', n_coils=1):
         """
         self.num_slices = shape[2]
         self.shape = shape
-        (kspace_plane_loc, self.z_sample_loc,
-         self.sort_pos) = get_stacks_fourier(kspace_loc)
+        (kspace_plane_loc, self.z_sample_loc, self.sort_pos, self.idx_mask_z) \
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: why can't we make this:
```python
(kspace_plane_loc, self.z_sample_loc, self.sort_pos, self.idx_mask_z) = get_stacks_fourier(
  kspace_loc,
  self.shape,
)
```
?

Comment (chaithyagr) [reply mri/operators/fourier/non_cartesian.py]: That would cross the PEP8 line length of 78 characters right?

Comment (zaccharieramzi) [reply mri/operators/fourier/non_cartesian.py]: https://www.python.org/dev/peps/pep-0008/#maximum-line-length
Since we are the only ones working on this project, I think it's fine to have a longer maximum line length like 99.

In the team I worked in before, because it was not an open-source project, we even used 120 characters.

File: mri/parallel_mri/extract_sensitivity_maps.py (hunk 10 above)
Comment (mathrip) [reply mri/parallel_mri/extract_sensitivity_maps.py]: Ok!

File: mri/operators/fourier/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mathrip) [reply mri/operators/fourier/utils.py]: Yes you are right. So I changed it in this function because I had this argument. But it exists also in other functions such as extract_sensitivity, and the class Stacked3DNFFT. I thought you wanted me to change the name here as well... but I was mistaken !

File: mri/operators/fourier/utils.py (hunk 26 above)
Comment (mathrip) [reply mri/operators/fourier/utils.py]: Ok, it will be on the next commit

File: mri/operators/fourier/utils.py (hunk 27 above)
Comment (mathrip) [reply mri/operators/fourier/utils.py]: done

File: mri/operators/fourier/non_cartesian.py (hunk 28 above)
Comment (mathrip) [reply mri/operators/fourier/non_cartesian.py]: It failed the test if I change it, because of pep8. I let it like this for now, but if you insist I'll change it.

File: mri/operators/fourier/utils.py (hunk 27 above)
Comment (mathrip) [reply mri/operators/fourier/utils.py]: Ok done

File: mri/operators/fourier/non_cartesian.py [hunk 29]
```diff
@@ -526,29 +537,26 @@ def op(self, data):
             coeff = [self._op(data[i])
                      for i in range(self.n_coils)]
         coeff = np.asarray(coeff)
+
```
[the remark this answers is not in the document]
Comment (mathrip) [reply mri/operators/fourier/non_cartesian.py]: Ok, it's done

**Comment by chaithyagr** [comment]
We may have to rebase this completely. This could be a hard task, based on changes we can as well take a new branch and do it manually. @mathrip do you want to do this? Or shall I take over?

**Comment by chaithyagr** [comment]
Taking over for rebasing and ensuring this is at mainline

**Comment by chaithyagr** [comment]
I have rebased all the codes to master. @zaccharieramzi can you please do a final check? 
I will merge once I have tests passing

**Comment by zaccharieramzi** [comment]
@mathrip have you had the chance to re-read the rebased version?

**Comment by mathrip** [comment]
> @mathrip have you had the chance to re-read the rebased version?

Not yet, but I'll do it tomorrow !

**Comment by mathrip** [comment]
> @mathrip have you had the chance to re-read the rebased version?

I reviewed and I approve @zaccharieramzi can you approve it? Thanks
